kinesis/kinesis.go

```diff
@@ -0,0 +1,330 @@
+    dataKeys                        string
+    // If specified, the value of that data key will be used as the partition key. 
+    // Otherwise a random string will be used.
+    // Partition key decides in which shard of your stream the data belongs to
+    partitionKey                    string
+    // Decides whether to append a newline after each data record
+    appendNewline                   bool
+    lastInvalidPartitionKeyIndex    int
+    client                          PutRecordsClient
+    records                         []*kinesis.PutRecordsRequestEntry
+    dataLength                      int
+    backoff                         *plugins.Backoff
+    timer                           *plugins.Timeout
+    PluginID                        int
+    random                          *random
+}
+
+// NewOutputPlugin creates an OutputPlugin object
+func NewOutputPlugin(region, stream, dataKeys, partitionKey, roleARN, endpoint string, appendNewline bool, pluginID int) (*OutputPlugin, error) {
+    client, err := newPutRecordsClient(roleARN, region, endpoint)
+    if err != nil {
+        return nil, err
+    }
+
+    records := make([]*kinesis.PutRecordsRequestEntry, 0, maximumRecordsPerPut)
+    timer, err := plugins.NewTimeout(func (d time.Duration) {
+        logrus.Errorf("[kinesis %d] timeout threshold reached: Failed to send logs for %s\n", pluginID, d.String())
+        logrus.Errorf("[kinesis %d] Quitting Fluent Bit", pluginID)
+        os.Exit(1)
+    })
+
+    if err != nil {
+        return nil, err
+    }
+
+    seededRand := rand.New(rand.NewSource(time.Now().UnixNano()))
+    random := &random{
+        seededRandom:   seededRand,
+        buffer:         make([]byte, 8),
+    }
+
+    return &OutputPlugin{
+        stream:                         stream,
+        client:                         client,
+        records:                        records,
+        dataKeys:                       dataKeys,
+        partitionKey:                   partitionKey,
+        appendNewline:                  appendNewline,
+        lastInvalidPartitionKeyIndex:   -1,
+        backoff:                        plugins.NewBackoff(),
+        timer:                          timer,
+        PluginID:                       pluginID,
+        random:                         random,
+    }, nil
+}
+
+// newPutRecordsClient creates the Kinesis client for calling the PutRecords method
+func newPutRecordsClient(roleARN string, awsRegion string, endpoint string) (*kinesis.Kinesis, error) {
+    sess, err := session.NewSession(&aws.Config{
+        Region: aws.String(awsRegion),
+    })
+    if err != nil {
+        return nil, err
+    }
+
+    svcConfig := &aws.Config{}
+    if endpoint != "" {
+        defaultResolver := endpoints.DefaultResolver()
+        cwCustomResolverFn := func(service, region string, optFns ...func(*endpoints.Options)) (endpoints.ResolvedEndpoint, error) {
+            if service == "kinesis" {
```

Comment:
minor: I didn't know this when I wrote the cloudwatch and firehose plugins, but there are SDK constants for the service names. 

There should be something like `endpoints.KinesisServiceID` so that you can do this:

```
if service == endpoints.KinesisServiceID
```

Ex: https://github.com/awslabs/amazon-ecs-local-container-endpoints/blob/master/local-container-endpoints/handlers/credentials_handler.go#L67